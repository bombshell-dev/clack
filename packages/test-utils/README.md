# `@clack/test-utils`

A bunch of useful test utiltiies for use inside clack.
