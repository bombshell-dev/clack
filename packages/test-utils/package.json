{
  "name": "@clack/test-utils",
  "private": true,
  "version": "1.0.0-alpha.1",
  "type": "module",
  "main": "./dist/index.js",
  "module": "./dist/index.js",
  "exports": {
    ".": {
      "types": "./dist/index.d.ts",
      "default": "./dist/index.js"
    },
    "./package.json": "./package.json"
  },
  "types": "./dist/index.d.ts",
  "repository": {
    "type": "git",
    "url": "git+https://github.com/bombshell-dev/clack.git",
    "directory": "packages/test-utils"
  },
  "bugs": {
    "url": "https://github.com/bombshell-dev/clack/issues"
  },
  "homepage": "https://github.com/bombshell-dev/clack/tree/main/packages/test-utils#readme",
  "files": ["dist", "CHANGELOG.md"],
  "author": {
    "name": "James Garbutt",
    "url": "https://github.com/43081j"
  },
  "license": "MIT",
  "packageManager": "pnpm@9.14.2",
  "scripts": {
    "build": "tsc",
    "prepack": "pnpm build"
  },
  "dependencies": {
  },
  "devDependencies": {
  }
}
