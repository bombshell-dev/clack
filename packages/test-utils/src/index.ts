export { MockReadable } from './mock-readable.js';
export { MockWritable } from './mock-writable.js';

export function nextTick(): Promise<void> {
	return new Promise<void>((resolve) => {
		setTimeout(resolve, 0);
	});
}
