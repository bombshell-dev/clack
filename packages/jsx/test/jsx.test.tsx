import { MockReadable, MockWritable } from '@clack/test-utils';
import { beforeEach, describe, expect, test } from 'vitest';
import { Confirm, Note, Option, Password, Select, Text, jsx } from '../src/index.js';

describe('jsx', () => {
	let input: MockReadable;
	let output: MockWritable;

	beforeEach(() => {
		input = new MockReadable();
		output = new MockWritable();
	});

	test('can render', async () => {
		const task = jsx(Confirm, {
			message: 'foo?',
			input,
			output,
		});
		input.emit('keypress', '', { name: 'return' });
		const result = await task;
		expect(result).to.equal(true);
	});

	test('can render JSX', async () => {
		const task = <Confirm message="foo?" input={input} output={output} />;
		input.emit('keypress', '', { name: 'return' });
		const result = await task;
		expect(result).to.equal(true);
	});

	test('unknown elements are null', async () => {
		const task = jsx('unknown-nonsense' as never, {} as never);
		const result = await task;
		expect(result).to.equal(null);
	});

	describe('Confirm', () => {
		test('can set message', async () => {
			const task = <Confirm message="foo?" input={input} output={output} />;
			input.emit('keypress', '', { name: 'return' });
			const result = await task;
			expect(result).to.equal(true);
			expect(output.buffer).toMatchSnapshot();
		});

		test('can set active text', async () => {
			const task = <Confirm message="foo?" active="DO IT" input={input} output={output} />;
			input.emit('keypress', '', { name: 'return' });
			const result = await task;
			expect(result).to.equal(true);
			expect(output.buffer).toMatchSnapshot();
		});

		test('can set inactive text', async () => {
			const task = <Confirm message="foo?" inactive="DONT DO IT" input={input} output={output} />;
			input.emit('keypress', '', { name: 'return' });
			const result = await task;
			expect(result).to.equal(true);
			expect(output.buffer).toMatchSnapshot();
		});
	});

	describe('Note', () => {
		test('can render string message', async () => {
			const task = <Note message="foo" output={output} />;
			await task;

			expect(output.buffer).toMatchSnapshot();
		});

		test('can render children as message', async () => {
			const task = <Note output={output}>a message</Note>;
			await task;

			expect(output.buffer).toMatchSnapshot();
		});

		test('can render complex results as message', async () => {
			const task = (
				<Note output={output}>
					<Confirm message="say yes" input={input} output={output} />
				</Note>
			);
			input.emit('keypress', '', { name: 'return' });
			await task;

			expect(output.buffer).toMatchSnapshot();
		});

		test('can render multiple children as message', async () => {
			const task = (
				<Note output={output}>
					<Confirm message="say yes" input={input} output={output} />
					<Confirm message="say yes again" input={input} output={output} />
				</Note>
			);
			input.emit('keypress', '', { name: 'return' });
			input.emit('keypress', '', { name: 'return' });
			await task;

			expect(output.buffer).toMatchSnapshot();
		});
	});

	describe('Text', () => {
		test('renders text input', async () => {
			const task = <Text message="foo" output={output} input={input} />;

			input.emit('keypress', '', { name: 'return' });

			const result = await task;

			expect(result).to.equal('');
			expect(output.buffer).toMatchSnapshot();
		});

		test('can set placeholder', async () => {
			const task = <Text message="foo" placeholder="bar" output={output} input={input} />;

			input.emit('keypress', '', { name: 'return' });

			const result = await task;

			expect(result).to.equal('');
			expect(output.buffer).toMatchSnapshot();
		});

		test('can set default value', async () => {
			const task = <Text message="foo" defaultValue="bar" output={output} input={input} />;

			input.emit('keypress', '', { name: 'return' });

			const result = await task;

			expect(result).to.equal('bar');
			expect(output.buffer).toMatchSnapshot();
		});

		test('can set initial value', async () => {
			const task = <Text message="foo" initialValue="bar" output={output} input={input} />;

			input.emit('keypress', '', { name: 'return' });

			const result = await task;

			expect(result).to.equal('bar');
			expect(output.buffer).toMatchSnapshot();
		});
	});

	describe('Password', () => {
		test('renders password input', async () => {
			const task = <Password message="foo" output={output} input={input} />;

			input.emit('keypress', '', { name: 'return' });

			const result = await task;

			expect(result).to.equal(undefined);
			expect(output.buffer).toMatchSnapshot();
		});

		test('renders user input', async () => {
			const task = <Password message="foo" output={output} input={input} />;

			input.emit('keypress', 'a', { name: 'a' });
			input.emit('keypress', 'b', { name: 'b' });
			input.emit('keypress', '', { name: 'return' });

			const result = await task;

			expect(result).to.equal('ab');
			expect(output.buffer).toMatchSnapshot();
		});

		test('can set custom mask', async () => {
			const task = <Password message="foo" mask="!" output={output} input={input} />;

			input.emit('keypress', 'a', { name: 'a' });
			input.emit('keypress', 'b', { name: 'b' });
			input.emit('keypress', '', { name: 'return' });

			const result = await task;

			expect(result).to.equal('ab');
			expect(output.buffer).toMatchSnapshot();
		});
	});

	describe('Select', () => {
		test('renders options', async () => {
			const task = (
				<Select message="pick one" input={input} output={output}>
					<Option value={303}>Option One</Option>
					<Option value={808}>Option Two</Option>
				</Select>
			);

			input.emit('keypress', '', { name: 'return' });

			const result = await task;

			expect(result).to.equal(303);
			expect(output.buffer).toMatchSnapshot();
		});
	});
});
