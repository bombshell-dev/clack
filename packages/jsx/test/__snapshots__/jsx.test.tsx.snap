// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`jsx > Confirm > can set active text 1`] = `
[
  "<cursor.hide>",
  "[90m│[39m
[36m◆[39m  foo?
[36m│[39m  [32m●[39m DO IT [2m/[22m [2m○[22m [2mNo[22m
[36m└[39m
",
  "<cursor.backward count=999><cursor.up count=4>",
  "<cursor.down count=1>",
  "<erase.down>",
  "[32m◇[39m  foo?
[90m│[39m  [2mDO IT[22m",
  "
",
  "<cursor.show>",
]
`;

exports[`jsx > Confirm > can set inactive text 1`] = `
[
  "<cursor.hide>",
  "[90m│[39m
[36m◆[39m  foo?
[36m│[39m  [32m●[39m Yes [2m/[22m [2m○[22m [2mDONT DO IT[22m
[36m└[39m
",
  "<cursor.backward count=999><cursor.up count=4>",
  "<cursor.down count=1>",
  "<erase.down>",
  "[32m◇[39m  foo?
[90m│[39m  [2mYes[22m",
  "
",
  "<cursor.show>",
]
`;

exports[`jsx > Confirm > can set message 1`] = `
[
  "<cursor.hide>",
  "[90m│[39m
[36m◆[39m  foo?
[36m│[39m  [32m●[39m Yes [2m/[22m [2m○[22m [2mNo[22m
[36m└[39m
",
  "<cursor.backward count=999><cursor.up count=4>",
  "<cursor.down count=1>",
  "<erase.down>",
  "[32m◇[39m  foo?
[90m│[39m  [2mYes[22m",
  "
",
  "<cursor.show>",
]
`;

exports[`jsx > Note > can render children as message 1`] = `
[
  "[90m│[39m
[32m◇[39m  [0m[0m [90m──────────╮[39m
[90m│[39m             [90m│[39m
[90m│[39m  [2ma message[22m  [90m│[39m
[90m│[39m             [90m│[39m
[90m├─────────────╯[39m
",
]
`;

exports[`jsx > Note > can render complex results as message 1`] = `
[
  "<cursor.hide>",
  "[90m│[39m
[36m◆[39m  say yes
[36m│[39m  [32m●[39m Yes [2m/[22m [2m○[22m [2mNo[22m
[36m└[39m
",
  "<cursor.backward count=999><cursor.up count=4>",
  "<cursor.down count=1>",
  "<erase.down>",
  "[32m◇[39m  say yes
[90m│[39m  [2mYes[22m",
  "
",
  "<cursor.show>",
  "[90m│[39m
[32m◇[39m  [0m[0m [90m─────╮[39m
[90m│[39m        [90m│[39m
[90m│[39m  [2mtrue[22m  [90m│[39m
[90m│[39m        [90m│[39m
[90m├────────╯[39m
",
]
`;

exports[`jsx > Note > can render multiple children as message 1`] = `
[
  "<cursor.hide>",
  "[90m│[39m
[36m◆[39m  say yes
[36m│[39m  [32m●[39m Yes [2m/[22m [2m○[22m [2mNo[22m
[36m└[39m
",
  "<cursor.hide>",
  "[90m│[39m
[36m◆[39m  say yes again
[36m│[39m  [32m●[39m Yes [2m/[22m [2m○[22m [2mNo[22m
[36m└[39m
",
  "<cursor.backward count=999><cursor.up count=4>",
  "<cursor.down count=1>",
  "<erase.down>",
  "[32m◇[39m  say yes
[90m│[39m  [2mYes[22m",
  "
",
  "<cursor.show>",
  "<cursor.backward count=999><cursor.up count=4>",
  "<cursor.down count=1>",
  "<erase.down>",
  "[32m◇[39m  say yes again
[90m│[39m  [2mYes[22m",
  "
",
  "<cursor.show>",
  "[90m│[39m
[32m◇[39m  [0m[0m [90m─────╮[39m
[90m│[39m        [90m│[39m
[90m│[39m  [2mtrue[22m  [90m│[39m
[90m│[39m  [2mtrue[22m  [90m│[39m
[90m│[39m        [90m│[39m
[90m├────────╯[39m
",
]
`;

exports[`jsx > Note > can render string message 1`] = `
[
  "[90m│[39m
[32m◇[39m  [0m[0m [90m────╮[39m
[90m│[39m       [90m│[39m
[90m│[39m  [2mfoo[22m  [90m│[39m
[90m│[39m       [90m│[39m
[90m├───────╯[39m
",
]
`;

exports[`jsx > Password > can set custom mask 1`] = `
[
  "<cursor.hide>",
  "[90m│[39m
[36m◆[39m  foo
[36m│[39m  [7m[8m_[28m[27m
[36m└[39m
",
  "<cursor.backward count=999><cursor.up count=4>",
  "<cursor.down count=2>",
  "<erase.line><cursor.left count=1>",
  "[36m│[39m  ![7m[8m_[28m[27m",
  "<cursor.down count=2>",
  "<cursor.backward count=999><cursor.up count=4>",
  "<cursor.down count=2>",
  "<erase.line><cursor.left count=1>",
  "[36m│[39m  !![7m[8m_[28m[27m",
  "<cursor.down count=2>",
  "<cursor.backward count=999><cursor.up count=4>",
  "<cursor.down count=1>",
  "<erase.down>",
  "[32m◇[39m  foo
[90m│[39m  [2m!![22m",
  "
",
  "<cursor.show>",
]
`;

exports[`jsx > Password > renders password input 1`] = `
[
  "<cursor.hide>",
  "[90m│[39m
[36m◆[39m  foo
[36m│[39m  [7m[8m_[28m[27m
[36m└[39m
",
  "<cursor.backward count=999><cursor.up count=4>",
  "<cursor.down count=1>",
  "<erase.down>",
  "[32m◇[39m  foo
[90m│[39m",
  "
",
  "<cursor.show>",
]
`;

exports[`jsx > Password > renders user input 1`] = `
[
  "<cursor.hide>",
  "[90m│[39m
[36m◆[39m  foo
[36m│[39m  [7m[8m_[28m[27m
[36m└[39m
",
  "<cursor.backward count=999><cursor.up count=4>",
  "<cursor.down count=2>",
  "<erase.line><cursor.left count=1>",
  "[36m│[39m  ▪[7m[8m_[28m[27m",
  "<cursor.down count=2>",
  "<cursor.backward count=999><cursor.up count=4>",
  "<cursor.down count=2>",
  "<erase.line><cursor.left count=1>",
  "[36m│[39m  ▪▪[7m[8m_[28m[27m",
  "<cursor.down count=2>",
  "<cursor.backward count=999><cursor.up count=4>",
  "<cursor.down count=1>",
  "<erase.down>",
  "[32m◇[39m  foo
[90m│[39m  [2m▪▪[22m",
  "
",
  "<cursor.show>",
]
`;

exports[`jsx > Text > can set default value 1`] = `
[
  "<cursor.hide>",
  "[90m│[39m
[36m◆[39m  foo
[36m│[39m  [7m[8m_[28m[27m
[36m└[39m
",
  "<cursor.backward count=999><cursor.up count=4>",
  "<cursor.down count=1>",
  "<erase.down>",
  "[32m◇[39m  foo
[90m│[39m  [2mbar[22m",
  "
",
  "<cursor.show>",
]
`;

exports[`jsx > Text > can set initial value 1`] = `
[
  "<cursor.hide>",
  "[90m│[39m
[36m◆[39m  foo
[36m│[39m  bar█
[36m└[39m
",
  "<cursor.backward count=999><cursor.up count=4>",
  "<cursor.down count=1>",
  "<erase.down>",
  "[32m◇[39m  foo
[90m│[39m  [2mbar[22m",
  "
",
  "<cursor.show>",
]
`;

exports[`jsx > Text > can set placeholder 1`] = `
[
  "<cursor.hide>",
  "[90m│[39m
[36m◆[39m  foo
[36m│[39m  [7mb[27m[2mar[22m
[36m└[39m
",
  "<cursor.backward count=999><cursor.up count=4>",
  "<cursor.down count=1>",
  "<erase.down>",
  "[32m◇[39m  foo
[90m│[39m",
  "
",
  "<cursor.show>",
]
`;

exports[`jsx > Text > renders text input 1`] = `
[
  "<cursor.hide>",
  "[90m│[39m
[36m◆[39m  foo
[36m│[39m  [7m[8m_[28m[27m
[36m└[39m
",
  "<cursor.backward count=999><cursor.up count=4>",
  "<cursor.down count=1>",
  "<erase.down>",
  "[32m◇[39m  foo
[90m│[39m",
  "
",
  "<cursor.show>",
]
`;
